primitives/src/circuit/merkle_tree/mod.rs

```diff
@@ -75,10 +77,133 @@ where
+    fn enforce_membership_proof(
         &mut self,
         elem: Self::LeafVar,
         merkle_proof: Self::MerklePathVar,
         expected_merkle_root: Variable,
     ) -> Result<(), CircuitError>;
 }
+
+/// Gadget for the sparse merkle tree
+pub trait SparseMerkleTreeGadget<M>: MerkleTreeGadget<M>
+where
+    M: MerkleTreeScheme,
+{
+    /// checking non-membership proof
+    fn is_non_member(
+        &mut self,
+        elem: Self::LeafVar,
+        merkle_proof: Self::MerklePathVar,
+        merkle_root: Variable,
+    ) -> Result<BoolVar, CircuitError>;
+
+    /// Enforce correct `merkle_proof` for the empty leaf `elem` against
+    /// `expected_merkle_root`.
+    fn enforce_non_membership_proof(
+        &mut self,
+        empty_elem: Self::LeafVar,
+        merkle_proof: Self::MerklePathVar,
+        expected_merkle_root: Variable,
+    ) -> Result<(), CircuitError>;
+}
+
+pub(crate) trait MerkleTreeHelperGadget<M>
+where
+    M: MerkleTreeScheme,
+{
+    type MembershipProofBooleanEncoding;
+    type MembershipProofVar;
+    /// Produces an ordered list of variables based on the relative position of
+    /// a node and its siblings.
+    /// * `node` - node to be inserted in the final list.
+    /// * `sibling1` - first sibling
+    /// * `sibling2` - second sibling
+    /// * `node_is_left` - variable that is true if node is the leftmost one.
+    /// * `node_is_right` -  variable that is true if node is the rightmost one.
+    /// * `returns` - list of variables corresponding to the node and its
+    ///   siblings in the correct order.
+    fn permute(
+        &mut self,
+        node: Variable,
+        sib1: Variable,
+        sib2: Variable,
+        node_is_left: BoolVar,
+        node_is_right: BoolVar,
+    ) -> Result<[Variable; 3], CircuitError>;
+
+    /// Ensure that the position of each node of the path is correctly encoded
+    /// Used for testing purposes.
+    /// * `merkle_path` - list of node of an authentication path
+    /// * `pos` - position of the missing leaf
+    /// * `returns` - list of variables corresponding to the authentication path
+    fn constrain_membership_proof(
+        &mut self,
+        merkle_path: &Self::MembershipProofBooleanEncoding,
+        pos: M::Index,
+    ) -> Result<Self::MembershipProofVar, CircuitError>;
+
+    /// Computes the merkle root based on some element placed at a leaf and a
+    /// merkle path.
+    /// * `elem` - variables corresponding to the uid and the element value
+    ///   (e.g.: record commitment).
+    /// * `path_vars` - variables corresponding to the Merkle path.
+    /// * `return` - variable corresponding to the root value of the Merkle
+    ///   tree.
+    fn compute_merkle_root(
+        &mut self,
+        elem: StandardLeafVar,
+        path_vars: &Self::MembershipProofVar,
+    ) -> Result<Variable, CircuitError>;
+}
+
+#[derive(Debug, Clone)]
+/// Circuit variable for a Merkle node.
+pub struct Rescue3AryNodeVar {
```

Comment:
Is it better to call it `Merkle3AryNodeVar`?

Reply:
Or `RescueMT3AryNodeVar`?

Reply:
I'd go with @chancharles92's suggestion here, since it's actually not Rescue specific. Thx!

Reply:
Done: https://github.com/EspressoSystems/jellyfish/pull/172/commits/86b0eb179b563747ad6099049e580056f02f5c61